src/Microsoft.AspNetCore.Server.Kestrel.Core/Internal/Http/Frame.cs

```diff
@@ -1420,5 +1446,57 @@ public void OnHeader(Span<byte> name, Span<byte> value)
 
             FrameRequestHeaders.Append(name, valueString);
         }
+
+        public void OnRequestParsingComplete()
+        {
+            ValidateHostHeader();
+        }
+
+        private void ValidateHostHeader()
+        {
+            if (_httpVersion == Http.HttpVersion.Http10)
+            {
+                return;
+            }
+
+            // https://tools.ietf.org/html/rfc7230#section-5.4
+            // A server MUST respond with a 400 (Bad Request) status code to any
+            // HTTP/1.1 request message that lacks a Host header field and to any
+            // request message that contains more than one Host header field or a
+            // Host header field with an invalid field-value.
+
+            var host = FrameRequestHeaders.HeaderHost;
+            if (host.Count <= 0)
+            {
+                RejectRequest(RequestRejectionReason.MissingHostHeader);
+            }
+            else if (host.Count > 1)
+            {
+                RejectRequest(RequestRejectionReason.MultipleHostHeaders);
+            }
+            else if (_requestTargetForm == HttpRequestTarget.AuthorityForm)
+            {
+                if (!host.Equals(RawTarget))
```

Comment:
Does this treat "www.example.com:80" and "www.example.com" as equivalent? Should it?